### ISSUE TITLE
Arithmetic comparisons

### ISSUE THREAD

**Issue by panos2310gr** [kyouko-taiga/LogicKit#15]
Is there a way to get results that satisfy arithmetic comparisons? e.g. 
kb.ask (.fact("number of children", var("x"), var("y")>3)

I suspect that built functions have this ability but I cannot find out a way to implement it.

Thanks.

**Comment by kyouko-taiga** [comment]
You would need to first convert your native Swift number into a term and implement "arithmetic" comparisons as LogicKit rules. There's indeed a built-in type and some helpers for that:

```swift
import LogicKit
import LogicKitBuiltins

let x = Nat.from(42)
let y = Nat.from(11)
let query =  Nat.greater(x, y)

let kb = KnowledgeBase(knowledge: Nat.relationAxioms)
kb.ask(query)
```

`Nat.relationAxioms` lets you "import" all the built-in definitions for LogicKit's natural numbers. Of course you will have to concatenate your own facts and rules to do anything interesting.

You can see some usage examples in the tests.

**Comment by panos2310gr** [comment]
Thank you.

Having populated the kb with my data I am trying to create a rule that takes Nat. Axioms into account:

var myrule = KnowledgeBase(knowledge: [
			Term.rule("high protein", .var("x"), .var("y")) {
			.fact("has protein value", .var("x"), .var("tx")) &&
			.fact("has protein value", .var("y"), .var("ty")) &&
				Nat.greater(Nat.asSwiftInt("x"),Nat.from(1))
		}])
		
It seems that I cannot get the syntax correct.
"Cannot convert value of type 'Int?' to expected argument type ‘Term’"

Any suggestions?

**Comment by kyouko-taiga** [comment]
TL;DR; Queries are built with terms and terms only. `Nat.asSwiftInt()`creates an optional `Int`, not a term. Just put your variable with `.var("x")` in `greater/2`.

Now for the longer explanation. There isn't any magic or special syntax behind builtins. They are simply helpers to create terms from some common native values. For instance, `Nat.from()` turns a Swift `Int` into a LogicKit `Term`. It is something like `nat::succ(nat::succ ...`. `Nat.asSwiftInt()` does the conversion in the other way. It takes a term (something built with `Nat.from()`) and gives you a Swift `Int`. The result is an optional, because statically LogicKit doesn't know if the term you specified is indeed a natural encoded the way it expects.

You can play with these helpers by printing the terms they produce.

**Comment by panos2310gr** [comment]
Ok so like this

var myrule = KnowledgeBase(knowledge: [
			Term.rule("high protein", .var("x"), .var("y")) {
			.fact("has protein value", .var("x"), .var("tx")) &&
			.fact("has protein value", .var("y"), .var("ty")) &&
			Nat.greater(.var(“y"),Nat.from(0))
		}])

kb = kb + myrule
		
		var answers = kb.ask (.fact("high protein", .var("a"), .var("b")))

		for answer in answers {
			print(answer)

		}
	


This does not return any values, although all y values are larger than 0

**Comment by kyouko-taiga** [comment]
I don't see anything in your rules to describe your `has protein value` predicate. So LogicKit is likely stuck at your first fact `.fact("has protein value", .var("x"), .var("tx"))`.

You can trace how LogicKit tries to solve your queries with the last argument of `KnowledgeBase.ask(:logger:)`.

```swift
var query = kb.ask(.fact("high protein", .var("a"), .var("b")), logger: DefaultLogger(useFontAttributes: false))
for ans in query.prefix(1) {
  print(ans)
}
```

This gives you the following:

```
Attempting to realize high protein[$a, $b]
using (high protein[$x, $y] ⊢ ((has protein value[$x, $tx] ∧ has protein value[$y, $ty]) ∧ nat::>[$y, nat::0])) 
Attempting to realize has protein value[$x, $tx]
```

**Comment by panos2310gr** [comment]
Sorry, complete code is:

for row in csvRows {

			let fact1 = Term.fact("has protein value",  .fact(row[0]), .var(row[1]))
			let fact2 = Term.fact("has fat value", .fact(row[0]),.var(row[2]))
			let fact3 = Term.fact("has carbohydrate value", .fact(row[0]),.fact(row[3]))
			let addedknowledge = [fact1,fact2,fact3]
			let kb1 = KnowledgeBase(knowledge:  addedknowledge)
			kb  = kb + kb1
		}
		
		let myrule = KnowledgeBase(knowledge: [
			Term.rule("high protein", .var("x"), .var("y")) {
			.fact("has protein value", .var("x"), .var("code")) &&
			.fact("has protein value", .var("y"), .var("content")) &&
			Nat.greater(.var("content"),Nat.from(0))
		}])
		
		kb = kb + myrule
		
		var answers = kb.ask (.fact("high protein", .var("a"), .var("b")))

		for answer in answers {
			print(answer)
		}
		
		answers = kb.ask (.fact("has protein value", .("a"), .var("b")))

		for answer in answers {
			print(answer)
		}

**Comment by kyouko-taiga** [comment]
Well, it still depends on what's coming from your CSV.

Without digging too much into the semantics of your application, my understanding of your rule is you need `y` to be unified with a non-zero natural number. I managed to get an answer that way:

```swift
var myrule = KnowledgeBase(knowledge: [
  Term.fact("has protein value", Nat.from(1), Nat.from(1)),
  Term.rule("high protein", .var("x"), .var("y")) {
    .fact("has protein value", .var("x"), .var("tx")) &&
      .fact("has protein value", .var("y"), .var("ty")) &&
      Nat.greater(.var("y"), Nat.from(0))
  }
])

let kb = KnowledgeBase(knowledge: Nat.arithmeticAxioms + myrule)
var query = kb.ask(.fact("high protein", .var("a"), .var("b")), logger: DefaultLogger(useFontAttributes: false))
for ans in query.prefix(1) {
  print(ans)
}
```

The answer in question:
```
["a": nat::succ[nat::0], "b": nat::succ[nat::0]]
```

I suspect there might be a bug with your predicates, or the way you built them. Try to inspect your KB (either with lldb or just adding some prints) and use the logger to spot them ;)

**Comment by panos2310gr** [comment]
Hi. I think I am getting close.

This is my code:

for row in csvRows {

			let fact1 = Term.fact("has protein value",  .fact(row[0]), .fact(row[1]))
			let fact2 = Term.fact("has fat value", .fact(row[0]),.fact(row[2]))
			let fact3 = Term.fact("has carbohydrate value", .fact(row[0]),.fact(row[3]))
			let fact4 = Term.fact("high protein", .fact(row[0]), .fact((row[4])))
			let fact5 = Term.fact("high fat", .fact(row[0]),.fact(row[5]))
			let fact6 = Term.fact("high glycemic", .fact(row[0]),.fact(row[6]))
			let fact7 = Term.fact("health index", .fact(row[0]), .fact(row[7]))
			let fact8 = Term.fact("age", .fact(row[0]), .lit(Int(row[8])))
			let fact9 = Term.fact("season", .fact(row[0]), .lit(Int(row[9])))
			let addedknowledge = [fact1,fact2,fact3,fact4,fact5,fact6,fact7,fact8,fact9]
			let kb1 = KnowledgeBase(knowledge:  addedknowledge)
			kb  = kb + kb1
		}
		
	
		let myrule = KnowledgeBase(knowledge: [
			.rule("is unhealthy", .var("x"), .var("y"), .var("z"), .var("season")) {
			.fact("high fat", .var("x"), "1") &&
			.fact("high glycemic", .var("x"), "1") &&
			.fact("health index", .var("x"), .var("y")) &&
			.fact("high protein", .var("x"), .var("z")) &&
			.fact("age", .var("x"), .var("agee")) &&
			Nat.greater(.var("agee"), Nat.from(5)) &&
			.fact("season", .var("x"), .var("season"))
		}])

		kb = kb + myrule
		
		let answers = kb.ask (.fact("is unhealthy", .var("x"), .var("y"),.var("z"),  .var("season")))
		for answer in answers {
			
			print(answer)
		}


Unfortunately, when I include the Nat.greater term in the rule, I get no results. If I exclude it I get results.

I am populating my kb from a csv and I want some integers in the kb, so that I can make comparisons. It seems that there are problems when converting from swift int to int for Nat library.

I understand that only natural numbers can be in Nat (1,2,3 etc, no zero, no negatives or fractions etc). There are some comparisons that I have to make in order to get specific results. 

Is this the correct way to do a comparison? What about selecting a specific integer value? Should I declare them as .lit?

regards.

**Comment by panos2310gr** [comment]
Also, this is the kb I construct:

[high fat[food_nutrient_conversion_factor_id, HighFat], high fat[16233, 0], high fat[16234, 0], high fat[16235, 0], high fat[16236, 0], high fat[16237, 0], high fat[16238, 0], high fat[16239, 0], high fat[16240, 0], high fat[16241, 0], high fat[16242, 0], high fat[16243, 0], high fat[16244, 0], high fat[16245, 0], high fat[16246, 0], high fat[16247, 0], high fat[16248, 0], high fat[16249, 0], high fat[16250, 0], high fat[16251, 1], high fat[16252, 1], high fat[16253, 1], high fat[16254, 1], high fat[16255, 1], high fat[16256, 1], high fat[16257, 1], high fat[16258, 1], high fat[16259, 1], high fat[16260, 1], high fat[16261, 1], high fat[16262, 1], high fat[16263, 1], high fat[16264, 1], high fat[16265, 1], high fat[16266, 1], high fat[16267, 1], high fat[16268, 1], high fat[16269, 1], high fat[16270, 1], high fat[16271, 1], high fat[16272, 1], high fat[16273, 1], high fat[16274, 1], high fat[16275, 1], high fat[16276, 1], high fat[16277, 1], high fat[16278, 1], high fat[16279, 1], high fat[16280, 1], high fat[16281, 1], high fat[16282, 1], high fat[16283, 0], high fat[16284, 0], high fat[16285, 0], high fat[16286, 0], high fat[16287, 0], high fat[16288, 0], high fat[16289, 0], high fat[16290, 0], high fat[16291, 0], high fat[16292, 0], high fat[16293, 0], high fat[16294, 0], nat::>[nat::succ[$x], nat::0], (nat::>[nat::succ[$x], nat::succ[$y]] ⊢ nat::>[$x, $y]), nat::<[nat::0, nat::succ[$x]], (nat::<[nat::succ[$x], nat::succ[$y]] ⊢ nat::<[$x, $y]), has protein value[food_nutrient_conversion_factor_id, protein_value], has protein value[16233, 4.27], has protein value[16234, 4.27], has protein value[16235, 4.27], has protein value[16236, 4.27], has protein value[16237, 4.27], has protein value[16238, 4.27], has protein value[16239, 4.27], has protein value[16240, 4.27], has protein value[16241, 4.27], has protein value[16242, 4.27], has protein value[16243, 4.27], has protein value[16244, 4.27], has protein value[16245, 4.27], has protein value[16246, 4.27], has protein value[16247, 4.27], has protein value[16248, 4.27], has protein value[16249, 4.27], has protein value[16250, 4.27], has protein value[16251, 4.27], has protein value[16252, 4.27], has protein value[16253, 4.27], has protein value[16254, 4.27], has protein value[16255, 4.27], has protein value[16256, 4.27], has protein value[16257, 4.27], has protein value[16258, 4.27], has protein value[16259, 4.27], has protein value[16260, 4.27], has protein value[16261, 4.27], has protein value[16262, 4.27], has protein value[16263, 4.27], has protein value[16264, 4.27], has protein value[16265, 4.27], has protein value[16266, 4.27], has protein value[16267, 4.27], has protein value[16268, 4.27], has protein value[16269, 4.27], has protein value[16270, 4.27], has protein value[16271, 4.27], has protein value[16272, 4.27], has protein value[16273, 4], has protein value[16274, 4], has protein value[16275, 3.9], has protein value[16276, 3.9], has protein value[16277, 3.47], has protein value[16278, 3.47], has protein value[16279, 3.47], has protein value[16280, 3.47], has protein value[16281, 3.47], has protein value[16282, 3.47], has protein value[16283, 3.47], has protein value[16284, 3.47], has protein value[16285, 3.47], has protein value[16286, 3.47], has protein value[16287, 3.47], has protein value[16288, 3.47], has protein value[16289, 3.47], has protein value[16290, 3.47], has protein value[16291, 3.47], has protein value[16292, 3.47], has protein value[16293, 3.47], has protein value[16294, 3.47], high glycemic[food_nutrient_conversion_factor_id, GlycemicHigh], high glycemic[16233, 1], high glycemic[16234, 1], high glycemic[16235, 1], high glycemic[16236, 1], high glycemic[16237, 1], high glycemic[16238, 1], high glycemic[16239, 1], high glycemic[16240, 1], high glycemic[16241, 1], high glycemic[16242, 1], high glycemic[16243, 1], high glycemic[16244, 1], high glycemic[16245, 1], high glycemic[16246, 1], high glycemic[16247, 0], high glycemic[16248, 0], high glycemic[16249, 0], high glycemic[16250, 0], high glycemic[16251, 1], high glycemic[16252, 1], high glycemic[16253, 1], high glycemic[16254, 0], high glycemic[16255, 0], high glycemic[16256, 0], high glycemic[16257, 0], high glycemic[16258, 0], high glycemic[16259, 0], high glycemic[16260, 0], high glycemic[16261, 0], high glycemic[16262, 0], high glycemic[16263, 0], high glycemic[16264, 0], high glycemic[16265, 0], high glycemic[16266, 0], high glycemic[16267, 0], high glycemic[16268, 0], high glycemic[16269, 0], high glycemic[16270, 0], high glycemic[16271, 0], high glycemic[16272, 0], high glycemic[16273, 0], high glycemic[16274, 0], high glycemic[16275, 0], high glycemic[16276, 0], high glycemic[16277, 0], high glycemic[16278, 0], high glycemic[16279, 0], high glycemic[16280, 0], high glycemic[16281, 0], high glycemic[16282, 0], high glycemic[16283, 0], high glycemic[16284, 0], high glycemic[16285, 0], high glycemic[16286, 0], high glycemic[16287, 0], high glycemic[16288, 0], high glycemic[16289, 0], high glycemic[16290, 0], high glycemic[16291, 0], high glycemic[16292, 0], high glycemic[16293, 0], high glycemic[16294, 0], has fat value[food_nutrient_conversion_factor_id, fat_value], has fat value[16233, 9.02], has fat value[16234, 9.02], has fat value[16235, 9.02], has fat value[16236, 9.02], has fat value[16237, 9], has fat value[16238, 9.02], has fat value[16239, 9.02], has fat value[16240, 9.02], has fat value[16241, 9.02], has fat value[16242, 9.02], has fat value[16243, 9.02], has fat value[16244, 9.02], has fat value[16245, 9.02], has fat value[16246, 9.02], has fat value[16247, 9.02], has fat value[16248, 9.02], has fat value[16249, 9.02], has fat value[16250, 9.02], has fat value[16251, 9.02], has fat value[16252, 9.02], has fat value[16253, 9.02], has fat value[16254, 9.02], has fat value[16255, 9], has fat value[16256, 9.02], has fat value[16257, 9.02], has fat value[16258, 9.02], has fat value[16259, 9.02], has fat value[16260, 9.02], has fat value[16261, 9.02], has fat value[16262, 9.02], has fat value[16263, 9.02], has fat value[16264, 9.02], has fat value[16265, 9.02], has fat value[16266, 9.02], has fat value[16267, 9.02], has fat value[16268, 9.02], has fat value[16269, 9.02], has fat value[16270, 9.02], has fat value[16271, 9.02], has fat value[16272, 9.02], has fat value[16273, 8.4], has fat value[16274, 9], has fat value[16275, 9], has fat value[16276, 8.8], has fat value[16277, 8.37], has fat value[16278, 8.37], has fat value[16279, 8.37], has fat value[16280, 8.37], has fat value[16281, 8.37], has fat value[16282, 8.37], has fat value[16283, 8.37], has fat value[16284, 8.37], has fat value[16285, 8.37], has fat value[16286, 8.37], has fat value[16287, 8.37], has fat value[16288, 8.37], has fat value[16289, 8.37], has fat value[16290, 8.37], has fat value[16291, 8.37], has fat value[16292, 8.37], has fat value[16293, 8.37], has fat value[16294, 8.37], high protein[food_nutrient_conversion_factor_id, HighProtein], high protein[16233, 1], high protein[16234, 1], high protein[16235, 1], high protein[16236, 1], high protein[16237, 1], high protein[16238, 1], high protein[16239, 1], high protein[16240, 1], high protein[16241, 1], high protein[16242, 1], high protein[16243, 1], high protein[16244, 1], high protein[16245, 1], high protein[16246, 1], high protein[16247, 1], high protein[16248, 1], high protein[16249, 1], high protein[16250, 1], high protein[16251, 1], high protein[16252, 0], high protein[16253, 1], high protein[16254, 1], high protein[16255, 1], high protein[16256, 0], high protein[16257, 0], high protein[16258, 0], high protein[16259, 0], high protein[16260, 0], high protein[16261, 0], high protein[16262, 0], high protein[16263, 0], high protein[16264, 0], high protein[16265, 0], high protein[16266, 0], high protein[16267, 0], high protein[16268, 0], high protein[16269, 0], high protein[16270, 0], high protein[16271, 0], high protein[16272, 0], high protein[16273, 0], high protein[16274, 0], high protein[16275, 0], high protein[16276, 0], high protein[16277, 0], high protein[16278, 0], high protein[16279, 0], high protein[16280, 0], high protein[16281, 0], high protein[16282, 0], high protein[16283, 0], high protein[16284, 0], high protein[16285, 0], high protein[16286, 0], high protein[16287, 0], high protein[16288, 0], high protein[16289, 0], high protein[16290, 0], high protein[16291, 0], high protein[16292, 0], high protein[16293, 0], high protein[16294, 0], age[food_nutrient_conversion_factor_id, nil], age[16233, 6], age[16234, 6], age[16235, 6], age[16236, 6], age[16237, 6], age[16238, 6], age[16239, 6], age[16240, 6], age[16241, 6], age[16242, 6], age[16243, 6], age[16244, 6], age[16245, 6], age[16246, 6], age[16247, 6], age[16248, 6], age[16249, 7], age[16250, 7], age[16251, 7], age[16252, 7], age[16253, 7], age[16254, 7], age[16255, 7], age[16256, 7], age[16257, 7], age[16258, 7], age[16259, 7], age[16260, 7], age[16261, 7], age[16262, 7], age[16263, 7], age[16264, 7], age[16265, 8], age[16266, 8], age[16267, 8], age[16268, 8], age[16269, 8], age[16270, 8], age[16271, 8], age[16272, 8], age[16273, 8], age[16274, 8], age[16275, 8], age[16276, 8], age[16277, 8], age[16278, 8], age[16279, 8], age[16280, 8], age[16281, 8], age[16282, 8], age[16283, 8], age[16284, 8], age[16285, 8], age[16286, 8], age[16287, 8], age[16288, 8], age[16289, 8], age[16290, 8], age[16291, 8], age[16292, 8], age[16293, 8], age[16294, 8], nat::<=[$x, $x], (nat::<=[$x, $y] ⊢ nat::<[$x, $y]), has carbohydrate value[food_nutrient_conversion_factor_id, carbohydrate_value], has carbohydrate value[16233, 3.87], has carbohydrate value[16234, 3.87], has carbohydrate value[16235, 3.87], has carbohydrate value[16236, 3.87], has carbohydrate value[16237, 3.87], has carbohydrate value[16238, 3.7], has carbohydrate value[16239, 3.87], has carbohydrate value[16240, 3.87], has carbohydrate value[16241, 3.87], has carbohydrate value[16242, 3.87], has carbohydrate value[16243, 3.87], has carbohydrate value[16244, 3.87], has carbohydrate value[16245, 3.87], has carbohydrate value[16246, 3.87], has carbohydrate value[16247, 3.87], has carbohydrate value[16248, 3.87], has carbohydrate value[16249, 3.87], has carbohydrate value[16250, 3.87], has carbohydrate value[16251, 3.87], has carbohydrate value[16252, 3.87], has carbohydrate value[16253, 3.87], has carbohydrate value[16254, 3.87], has carbohydrate value[16255, 3.87], has carbohydrate value[16256, 3.87], has carbohydrate value[16257, 3.87], has carbohydrate value[16258, 3.87], has carbohydrate value[16259, 3.87], has carbohydrate value[16260, 3.87], has carbohydrate value[16261, 3.87], has carbohydrate value[16262, 4.11], has carbohydrate value[16263, 3.87], has carbohydrate value[16264, 3.87], has carbohydrate value[16265, 3.87], has carbohydrate value[16266, 3.87], has carbohydrate value[16267, 4.11], has carbohydrate value[16268, 4.11], has carbohydrate value[16269, 4.11], has carbohydrate value[16270, 4.11], has carbohydrate value[16271, 4.11], has carbohydrate value[16272, 3.87], has carbohydrate value[16273, 3.4], has carbohydrate value[16274, 3.8], has carbohydrate value[16275, 3.7], has carbohydrate value[16276, 3.6], has carbohydrate value[16277, 4.07], has carbohydrate value[16278, 4.07], has carbohydrate value[16279, 4.07], has carbohydrate value[16280, 4.07], has carbohydrate value[16281, 4.07], has carbohydrate value[16282, 4.07], has carbohydrate value[16283, 4.07], has carbohydrate value[16284, 4.07], has carbohydrate value[16285, 4.07], has carbohydrate value[16286, 4.07], has carbohydrate value[16287, 4.07], has carbohydrate value[16288, 4.07], has carbohydrate value[16289, 4.07], has carbohydrate value[16290, 4.07], has carbohydrate value[16291, 4.07], has carbohydrate value[16292, 4.07], has carbohydrate value[16293, 4.07], has carbohydrate value[16294, 4.07], season[food_nutrient_conversion_factor_id, nil], season[16233, 4], season[16234, 4], season[16235, 4], season[16236, 4], season[16237, 4], season[16238, 4], season[16239, 4], season[16240, 4], season[16241, 4], season[16242, 4], season[16243, 4], season[16244, 4], season[16245, 4], season[16246, 4], season[16247, 4], season[16248, 4], season[16249, 4], season[16250, 4], season[16251, 4], season[16252, 4], season[16253, 4], season[16254, 2], season[16255, 2], season[16256, 2], season[16257, 2], season[16258, 2], season[16259, 2], season[16260, 2], season[16261, 2], season[16262, 2], season[16263, 2], season[16264, 2], season[16265, 2], season[16266, 2], season[16267, 2], season[16268, 2], season[16269, 2], season[16270, 2], season[16271, 3], season[16272, 3], season[16273, 3], season[16274, 3], season[16275, 3], season[16276, 3], season[16277, 3], season[16278, 3], season[16279, 3], season[16280, 3], season[16281, 3], season[16282, 3], season[16283, 3], season[16284, 3], season[16285, 1], season[16286, 1], season[16287, 1], season[16288, 1], season[16289, 1], season[16290, 1], season[16291, 1], season[16292, 1], season[16293, 1], season[16294, 1], (is unhealthy[$x, $y, $z, $season] ⊢ ((((((high fat[$x, 1] ∧ high glycemic[$x, 1]) ∧ health index[$x, $y]) ∧ high protein[$x, $z]) ∧ age[$x, $agee]) ∧ nat::>[$agee, nat::succ[nat::succ[nat::succ[nat::succ[nat::succ[nat::0]]]]]]) ∧ season[$x, $season])), health index[food_nutrient_conversion_factor_id, HealthIndex], health index[16233, 1], health index[16234, 1], health index[16235, 1], health index[16236, 1], health index[16237, 1], health index[16238, 1], health index[16239, 1], health index[16240, 1], health index[16241, 1], health index[16242, 1], health index[16243, 1], health index[16244, 1], health index[16245, 2], health index[16246, 2], health index[16247, 2], health index[16248, 2], health index[16249, 2], health index[16250, 2], health index[16251, 5], health index[16252, 4], health index[16253, 3], health index[16254, 2], health index[16255, 2], health index[16256, 2], health index[16257, 2], health index[16258, 2], health index[16259, 3], health index[16260, 3], health index[16261, 3], health index[16262, 3], health index[16263, 3], health index[16264, 3], health index[16265, 3], health index[16266, 3], health index[16267, 3], health index[16268, 3], health index[16269, 4], health index[16270, 4], health index[16271, 4], health index[16272, 4], health index[16273, 4], health index[16274, 4], health index[16275, 4], health index[16276, 4], health index[16277, 5], health index[16278, 5], health index[16279, 5], health index[16280, 5], health index[16281, 5], health index[16282, 5], health index[16283, 5], health index[16284, 5], health index[16285, 5], health index[16286, 5], health index[16287, 5], health index[16288, 5], health index[16289, 5], health index[16290, 5], health index[16291, 5], health index[16292, 5], health index[16293, 5], health index[16294, 5], nat::>=[$x, $x], (nat::>=[$x, $y] ⊢ nat::>[$x, $y])]

**Comment by kyouko-taiga** [comment]
> I understand that only natural numbers can be in Nat (1,2,3 etc, no zero, no negatives or fractions etc). There are some comparisons that I have to make in order to get specific results.

Natural numbers include 0, but you are right for the other numbers.

> I am populating my kb from a csv and I want some integers in the kb, so that I can make comparisons. It seems that there are problems when converting from swift int to int for Nat library.

Your CSV is probably giving you strings, with a number. I see that you wrote this:

```
let fact1 = Term.fact("has protein value",  .fact(row[0]), .fact(row[1]))
```

If `row` is, as I suspect, an array of strings, then it won't work. While you _see_ `16266` when you print, it's not a number that has been stored in your variables. You need to:
1. Convert your strings into native swift integers (simply, `let i = Int("12345")!`)
2. Convert your swift integers to LogicKit naturals (simply, `Nat.from(i)`)

> Is this the correct way to do a comparison? What about selecting a specific integer value? Should I declare them as .lit?

This construct is just a way to wrap a native value as an opaque object in LogicKit. You won't be able to do any sort of comparison with a literal, unless you manually enumerate the relationships. This is not a valid use case in your situation.

**Comment by panos2310gr** [comment]
Dear Dimitri,

I have managed to load my data in a knowledgebase and I would like to get the number of items in an answerset. Underestimatedcount returns 0.

Is there a way to do this without adding counters in my loop to display answers?

Regards

**Comment by kyouko-taiga** [comment]
TL;DR; You can try enumerating all the answer set in an array and use `Array.count` afterwards. There's no way to get the total number of elements without enumerating the entire answer set. `underestimatedCount` is not a reliable property to get the number of elements in a sequence, as it should only return the number of elements a type can compute non-destructively in O(1) (see [Swift's documentation](https://developer.apple.com/documentation/swift/sequence/3018375-underestimatedcount)).

Now, I wouldn't recommend this unless you are _sure_ your program will always terminate after finding a finite number of answers. Otherwise, you risk getting stuck with your interpreter trying to find new solutions until the end of times. This can happen if your program contains a rule that creates a larger (or more complex) term to unify its head. Consider for example:

```swift
let zero: Term = "zero"
let x: Term = .var("x")
let y: Term = .var("y")
let z: Term = .var("z")

let kb: KnowledgeBase = [
   .fact("add", zero, y, y),
   .fact("add", .fact("succ", x), y, z) |-
     .fact("add", x, .fact("succ", y), z),
]

var answers = kb.ask(.fact("add", x, y, .fact("succ", .fact("succ", zero))))

// This never terminates :(
print(Array(answers).count)
```

The problem here is that once you've computed `0 + 2`, `1 + 1` and `2 + 0`, a logic program interpreter will try to find `y` such that `3 + y = 2` and obviously never succeed doing so.

**Comment by panos2310gr** [comment]
I found this way to be helpful

 let answers2 = kb.ask (.fact( "fatTotalOrder" , .var("x"), "1"))
    print (answers2.all.count)

Is this safe and accurate?

Thanks!

**Comment by kyouko-taiga** [comment]
I guess I wrote `AnswerSet.all` as a syntactic sugar for the call to the the array initializer. Both achieve exactly the same, as you can see from the source: https://github.com/kyouko-taiga/LogicKit/blob/3e32454479684da7f37f8fe31c73724735b71d04/Sources/LogicKit/AnswerSet.swift#L15

**Comment by panos2310gr** [comment]
Thank you!

This seems to be destructive though, as you mention. All queries after this (eg. Next()) return nil, as the array is run through). Same happens if I run something like  print (answers.contains(["x" : "4"])), the element is removed from the array. Did I get this right? Is this what is meant with “destructive”? 

Is there a way to reset the answers to the result right after the kb.ask?

Also, I cannot seem to find a way to get a answers.max syntax correctly.

If I understand it, it will result with the value that is the maximum (as a Nat probably?) in a given result set.

**Comment by kyouko-taiga** [comment]
There's nothing fancy about `AnswerSet`, it's just a sequence. Everything that holds on other Swift sequences (e.g., [`JoinedSequence`](https://developer.apple.com/documentation/swift/joinedsequence), [`Zip2Sequence`](https://developer.apple.com/documentation/swift/zip2sequence), ...) also holds on `AnswerSet`.

If you build an array (or any other collection), then iterating over the elements or checking for containment will not be destructive anymore. So if you call `AnswerSet.all` or if you put all your binding maps in say a `Set`, then you can do whatever you want with them. [Swift's documentation](https://developer.apple.com/documentation/swift/swift_standard_library/collections/sequence_and_collection_protocols) will give you a very clear explanation on this.

As `AnswerSet` is a sequence, it gets a `max(by:)` method that returns its "maximum" element. But unless your elements conform to `Comparable`, Swift has no way to determine the maximum element, so the method accepts a closure that does just that. Once again, I refer you to [Swift's documentation](https://developer.apple.com/documentation/swift/array/2294243-max).

You can always re-query the same thing again to get another "rewinded" answer set. They are merely iterators that are initialized by `KnowledgeBase.ask(_:)`.

```swift
// Prints "["x": 0, "y": 2]" (or something equivalent)
var a0 = kb.ask(.fact("add", x, y, .fact("succ", .fact("succ", zero))))
print(a0.next())

// Prints "["x": 0, "y": 2]" (or something equivalent)
var a1 = kb.ask(.fact("add", x, y, .fact("succ", .fact("succ", zero))))
print(a1.next())

// Prints "["x": 1, "y": 1]" (or something equivalent)
print(a0.next())
```

This thread has strayed very far from the original question and is slowly turning into a Q/A on Swift sequences, so I'll close it. Feel free to open a different issue if you have another question specific to LogicKit.
